theano/sandbox/gpuarray/basic_ops.py

```diff
@@ -19,28 +20,49 @@
 except ImportError:
     pass
 
-from type import GpuArrayType
+from .type import GpuArrayType, gpu_context_type, get_context
 
+# This is a marker to indicate that any context is acceptable
+# It should never be registered as a context name
+AnyContext = object()
 
-def as_gpuarray_variable(x):
+def as_gpuarray_variable(x, context):
     # This is needed to lower the number of useless transfer
     # introduced during optimization.  This speed up optimization and
     # "canonicalize" the graph, so it make easier making some
     # optimization.
-    if (hasattr(x, 'fgraph') and
-        len(x.clients) == 1 and
-        x.owner and
-        isinstance(x.owner.op, HostFromGpu)):
-        return x.owner.inputs[0]
     if hasattr(x, '_as_GpuArrayVariable'):
-        return x._as_GpuArrayVariable()
-    # TODO we need to have the cuda -> gpu path taken care of.
-    tensor_x = tensor.as_tensor_variable(x)
-    return gpu_from_host(tensor_x)
+        return x._as_GpuArrayVariable(context)
+
+    x = tensor.as_tensor_variable(x)
+
+    if (x.owner and
+        isinstance(x.owner.op, HostFromGpu) and
+        context == x.owner.inputs[0].type.context):
+        return x.owner.inputs[0]
 
+    return GpuFromHost(context)(x)
 
-def as_gpuarray(x):
-    return gpuarray.array(x, copy=False)
+
+def infer_context(*vars):
+    "Infer the context to use from the inputs given"
```

Comment:
I can try to do that but I won't really have the time to go through the code this week to make this change.

I'm not sure I like "context name", but a different term to indicate the difference in meaning is a good idea.